akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedBehavior.scala

```diff
@@ -0,0 +1,102 @@
+/**
+ * Copyright (C) 2016-2018 Lightbend Inc. <http://www.lightbend.com/>
+ */
+
+package akka.persistence.typed.internal
+
+import java.util.UUID
+import java.util.concurrent.atomic.AtomicInteger
+
+import akka.actor.NoSerializationVerificationNeeded
+import akka.actor.typed.Behavior
+import akka.actor.typed.Behavior.StoppedBehavior
+import akka.actor.typed.scaladsl.{ ActorContext, TimerScheduler }
+import akka.annotation.InternalApi
+import akka.event.{ LogSource, Logging }
+import akka.persistence.typed.scaladsl.PersistentBehaviors
+import akka.persistence.{ JournalProtocol, Persistence, RecoveryPermitter, SnapshotProtocol }
+import akka.{ actor ⇒ a }
+
+/** INTERNAL API */
+@InternalApi
+private[akka] object EventsourcedBehavior {
+
+  // ok to wrap around (2*Int.MaxValue restarts will not happen within a journal roundtrip)
+  private[akka] val instanceIdCounter = new AtomicInteger(1)
+
+  @InternalApi private[akka] object WriterIdentity {
+    def newIdentity(): WriterIdentity = {
+      val instanceId: Int = EventsourcedBehavior.instanceIdCounter.getAndIncrement()
+      val writerUuid: String = UUID.randomUUID.toString
+      WriterIdentity(instanceId, writerUuid)
+    }
+  }
+  private[akka] final case class WriterIdentity(instanceId: Int, writerUuid: String)
+
+  /** Protocol used internally by the eventsourced behaviors, never exposed to user-land */
+  private[akka] sealed trait EventsourcedProtocol
+  private[akka] case object RecoveryPermitGranted extends EventsourcedProtocol
+  private[akka] final case class JournalResponse(msg: akka.persistence.JournalProtocol.Response) extends EventsourcedProtocol
+  private[akka] final case class SnapshotterResponse(msg: akka.persistence.SnapshotProtocol.Response) extends EventsourcedProtocol
+  private[akka] final case class RecoveryTickEvent(snapshot: Boolean) extends EventsourcedProtocol
+  private[akka] final case class ReceiveTimeout(timeout: akka.actor.ReceiveTimeout) extends EventsourcedProtocol
+
+  implicit object PersistentBehaviorLogSource extends LogSource[EventsourcedBehavior[_, _, _]] {
+    override def genString(b: EventsourcedBehavior[_, _, _]): String = {
+      val behaviorShortName = b match {
+        case _: EventsourcedRunning[_, _, _]                  ⇒ "running"
+        case _: EventsourcedRecoveringEvents[_, _, _]         ⇒ "recover-events"
+        case _: EventsourcedRecoveringSnapshot[_, _, _]       ⇒ "recover-snap"
+        case _: EventsourcedRequestingRecoveryPermit[_, _, _] ⇒ "awaiting-permit"
+      }
+      s"PersistentBehavior[id:${b.persistenceId}][${b.context.self.path}][$behaviorShortName]"
+    }
+  }
```

Comment:
Hmm, while neat, is this really needed, can't we just use `ctx.log` rather than separate logging instances in the four behaviors? Maybe put the `id` and `behaviorShortName` in mdc?